### PR TITLE
xds: improve error message when matched route on client is not of type RouteActionRoute

### LINKED ISSUE #5921: xds: report more descriptive error when matched route on client is not of type `RouteActionRoute`
In the config selector's `SelectConfig` method, we look for a matching route, and if we don't find one or the matched route does not contain clusters, we fail the RPC: https://github.com/grpc/grpc-go/blob/f2fbb0e07ebf3dd46aa641ee89c9f17e8083eaf6/xds/internal/resolver/serviceconfig.go#L158

It would be better if we check the route action type on the matched route and return a more meaningful error, when the route action type is not `RouteActionRoute` which is the only supported one on the client. https://github.com/grpc/grpc-go/blob/master/xds/internal/xdsclient/xdsresource/type_rds.go#L101-L115

### PR DESCRIPTION
**Pull request by Aditya-Sood**
Fixes #5921

RELEASE NOTES: none

### CODE REVIEW

**Comment by Aditya-Sood** [comment]
I've submitted the CLA document

File: xds/internal/resolver/serviceconfig.go [hunk 1]
```diff
@@ -156,7 +159,16 @@ func (cs *configSelector) SelectConfig(rpcInfo iresolver.RPCInfo) (*iresolver.RP
...
+		var err error
+		switch rt.actionType {
+		case xdsresource.RouteActionUnsupported:
+			err = errRouteActionUnsupported
+		case xdsresource.RouteActionNonForwardingAction:
+			err = errRouteActionNonForwardingAction
+		default:
+			err = errNoMatchedRouteFound
+		}
+		return nil, err
```
Comment (tobotg) [inline xds/internal/resolver/serviceconfig.go:171]: What about returning 
- `errNoMatchedRouteFound` error if `rt.clusters` is `nil` as it was done before this fix?
- and error as instructed in the [comment](https://github.com/grpc/grpc-go/issues/5921#issue-1528119137) only if route action type is not `RouteActionRoute`? And maybe format the error like `status.Errorf(codes.Unavailable, "received route was of type %t, require type RouteActionRoute", rt.actionType)`

Comment (Aditya-Sood) [reply xds/internal/resolver/serviceconfig.go:171]: hi @tobotg, thank you for reviewing!
1. > errNoMatchedRouteFound error if rt.clusters is nil as it was done before this fix?

for `rt.clusters == nil && rt.actionType == xdsresource.RouteActionRoute`,  the default case of the switch will handle it same as before (`return nil, errNoMatchedRouteFound`) - is that satisfactory or did I misunderstand what you were saying?

2. > and error as instructed in the [comment](https://github.com/grpc/grpc-go/issues/5921#issue-1528119137) only if route action type is not RouteActionRoute

do you mean that the `rt.clusters == nil` condition is not required for routes other than `RouteActionRoute`,  and so it should be a separate check on route type only?
@easwars could you please confirm?
if this is the case then I'll rewrite the block as follows:
 ```
if rt == nil {
	return nil, errNoMatchedRouteFound
}

if rt.actionType != xdsresource.RouteActionRoute {
	return nil, status.Errorf(codes.Unavailable, "received route was of type %t, require type RouteActionRoute", rt.actionType)
}

if rt.clusters == nil {
	return nil, errNoMatchedRouteFound
}
```

File: xds/internal/resolver/serviceconfig.go [hunk 2]
```diff
@@ -156,7 +159,16 @@ func (cs *configSelector) SelectConfig(rpcInfo iresolver.RPCInfo) (*iresolver.RP
 		}
 	}
 	if rt == nil || rt.clusters == nil {
-		return nil, errNoMatchedRouteFound
+		var err error
+		switch rt.actionType {
```
[the remark this answers is not in the document]
Comment (Aditya-Sood) [reply xds/internal/resolver/serviceconfig.go:163]: thanks for pointing out the bug!
I'll include it along with the fix that is decided for the next comment

File: xds/internal/resolver/serviceconfig.go [hunk 3]
```diff
@@ -121,6 +121,7 @@ type routeCluster struct {
 
 type route struct {
 	m                 *xdsresource.CompositeMatcher // converted from route matchers
+	actionType        xdsresource.RouteActionType   //holds route action type from type_rds.Route
```
Comment (easwars) [inline xds/internal/resolver/serviceconfig.go:124]:
```suggestion
	actionType        xdsresource.RouteActionType   //holds route action type 
```

File: xds/internal/resolver/serviceconfig.go [hunk 4]
```diff
@@ -142,6 +143,8 @@ type configSelector struct {
 }
 
 var errNoMatchedRouteFound = status.Errorf(codes.Unavailable, "no matched route was found")
+var errRouteActionUnsupported = status.Errorf(codes.Unavailable, "received route was of type RouteActionUnsupported, require type RouteActionRoute")
+var errRouteActionNonForwardingAction = status.Errorf(codes.Unavailable, "received route was of type RouteActionNonForwardingAction, require type RouteActionRoute")
```
Comment (easwars) [inline xds/internal/resolver/serviceconfig.go:147]: I would recommend getting rid of all these error sentinals and inlining them in the code.

Also, the error string for the errors added as part of this PR should be more like:
`matched route's action is type %s, only supported type is %s`.

You would also have to add a `String()` method to the `xdsresource.RouteActionType` enum, to satisfy the `Stringer` interface. See [this](https://go.dev/tour/methods/17) for details on the `Stringer` interface.

File: xds/internal/resolver/serviceconfig.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (easwars) [reply xds/internal/resolver/serviceconfig.go:163]: Yes, @tobotg is right here.

File: xds/internal/resolver/serviceconfig.go (hunk 1 above)
Comment (easwars) [reply xds/internal/resolver/serviceconfig.go:171]: Yes, you are right in the sense that for route with action type which is not `RouteActionRoute`, the `clusters` field will be `nil`. So, we will have to handle all three cases separately, as mentioned in your reply here.

**Comment by easwars** [comment]
@Aditya-Sood 

> Could you give me some pointers on how I can test this code? I've run the go test tests but the vet script was throwing a command not found error which I'm working on debugging

`go test google.golang.org/grpc/...` from the grpc-go directory should run all the tests. Also, you should add a test for RPCs matching routes with action not equal to `RouteActionRoute` and ensure that RPCs fail with the appropriate error. There are bunch of tests in https://github.com/grpc/grpc-go/blob/master/xds/internal/resolver/xds_resolver_test.go which verify different scenarios, and you should be able to a new one for this scenario. Please let me know if you have more questions on the test.

**Review by easwars — CHANGES_REQUESTED** [review]

**Comment by Aditya-Sood** [comment]
thanks for the review @easwars!

I've locally implemented all the changes that have been suggested

they are breaking a couple of tests in xds_resolver_test.go, so I'll work on resolving those + adding the `routes with action not equal to RouteActionRoute` test case before creating a commit and pushing it to the PR

**Comment by Aditya-Sood** [comment]
hi @easwars and @tobotg 
it seems like there are testcases that use `RouteActionUnsupported` routes on xds, which are now failing because of this additional check on route type:
```
1. Test/XDSResolverHTTPFilters/no_router_filter
-> Expects "no router filter present" from newInterceptor()
-> Receiving "action type of matched route is RouteActionUnsupported"

2. Test/XDSResolverHTTPFilters/ignored_after_router_filter
-> Expects err to be nil (throwing "Unexpected error from cs.SelectConfig(_)")
-> Receiving "action type of matched route is RouteActionUnsupported"

3. Test/XDSResolverHTTPFilters/NewStream_error
-> Expects err to be nil (throwing "Unexpected error from cs.SelectConfig(_)")
-> Receiving "action type of matched route is RouteActionUnsupported"

4. Test/XDSResolverHTTPFilters/all_overrides
-> Expects err to be nil (throwing "Unexpected error from cs.SelectConfig(_)")
-> Receiving "action type of matched route is RouteActionUnsupported"
```

I also tried flipping the order and putting the `if rt.clusters == nil {...}` statement  before the `if rt.actionType != xdsresource.RouteActionRoute {...}` statement, and the tests are still failing

(Output of `go test google.golang.org/grpc/...` - [test4.txt](https://github.com/grpc/grpc-go/files/11405515/test4.txt))

does this contradict our working assumption that `route with action type which is not RouteActionRoute, the clusters field will be nil`? or are these valid tests which now need to be modified to expect an error `if rt.actionType != RouteActionRoute`?

**Comment by easwars** [comment]
@Aditya-Sood 
Thank you so much for bringing up the issue of the failing tests. We had a discussion about this, and the decision is to ensure that an incoming LDS response is NACK'ed at the xDS client *if* the list of HTTP filters does not contain the router filter and/or the router filter is not the last one in the list.

Would you be interested in making that change? Once we do that, we can stop checking that the router filter is the last one in the list here.

I can provide you with more information if you are interested to make that change. Please let me know.

**Comment by Aditya-Sood** [comment]
hi @easwars, yes I can work on that change too, please share the details with me

**Comment by easwars** [comment]
> hi @easwars, yes I can work on that change too, please share the details with me

Here you go: https://github.com/grpc/grpc-go/issues/6259.

**Comment by easwars** [comment]
I'm setting the `Blocked` label on this one since we need to fix #6259 before getting back to this. Thanks for your help @Aditya-Sood

File: xds/internal/resolver/serviceconfig.go [hunk 5]
```diff
@@ -142,6 +143,8 @@ type configSelector struct {
 }
 
 var errNoMatchedRouteFound = status.Errorf(codes.Unavailable, "no matched route was found")
+var errRouteActionUnsupported = status.Errorf(codes.Unavailable, "received route was of type RouteActionUnsupported, require type RouteActionRoute")
```
Comment (dfawley) [inline xds/internal/resolver/serviceconfig.go:146]: These types are not the types that are configured in xDS, and aren't right to be using in the error messages here.

If we want to error with the proper type, we need to either define a `xdsresource.RouteActionType` for every xds action type, or we need to plumb in the actual xds route action enum.

File: xds/internal/resolver/serviceconfig.go (hunk 3 above)
Comment (zasweq) [reply xds/internal/resolver/serviceconfig.go:124]: Nit: also please add a space in between // and comment.

**Comment by zasweq** [comment]
So newInterceptor() shouldn't be checking for Router Filter and ignoring HTTP filters after, this was the legacy behavior that I should've deleted (the language iterated on here and deleted: https://github.com/grpc/proposal/pull/250/files) when I added the NACK check in the client for IsTerminal() here: https://github.com/grpc/grpc-go/pull/4676. newInterceptor() should simply just run the RPCs through all the filters.

File: xds/internal/resolver/serviceconfig.go (hunk 5 above)
Comment (Aditya-Sood) [reply xds/internal/resolver/serviceconfig.go:146]: hi @dfawley 
do you mean that instead of a generic `RouteActionUnsupported` we need to return a more precise `RouteActionType`? 
if so could you guide me on how to identify the other possible xds route actions?

cc @easwars

**Comment by Aditya-Sood** [comment]
hi @zasweq 
do you mean we should delete this section of code in newInterceptor()?
```
if router.IsRouterFilter(filter.Filter) {
	// Ignore any filters after the router filter.  The router itself
	// is currently a nop.
	return &interceptorList{interceptors: interceptors}, nil
}
```
([link to serviceconfig.go file](https://github.com/grpc/grpc-go/blob/2ee1fdd849505165744186e0cbe6ec07d6709514/xds/internal/resolver/serviceconfig.go#L295-L299))

**Comment by zasweq** [comment]
Yeah, and also any check for the router at all. Once it gets there, the scope is just run the RPC through all the filters.

**Comment by Aditya-Sood** [comment]
hi @zasweq 
so the `newInterceptor()` method in [serviceconfig.go](https://github.com/grpc/grpc-go/blob/2ee1fdd849505165744186e0cbe6ec07d6709514/xds/internal/resolver/serviceconfig.go#L294-L320) is the only location where the `router.IsRouterFilter()` check is currently being used

the method iterates over the `ConfigSelector`'s filters, building a client interceptor for each of them and appending them to a slice

the loop stops either when:
1. it encounters a router filter, returning the slice of client-interceptors; or 
2. all filters have been covered, at which point it returns an error stating "no router filter was found"

considering we are removing the router-filter check, should that loop now simply return the interceptors slice after it finishes iterating? something like so:
```
for _, filter := range cs.httpFilterConfig {
    override := cluster.httpFilterConfigOverride[filter.Name] // cluster is highest priority
    if override == nil {
        override = rt.httpFilterConfigOverride[filter.Name] // route is second priority
    }
    if override == nil {
        override = cs.virtualHost.httpFilterConfigOverride[filter.Name] // VH is third & lowest priority
    }
    ib, ok := filter.Filter.(httpfilter.ClientInterceptorBuilder)
    if !ok {
        // Should not happen if it passed xdsClient validation.
        return nil, fmt.Errorf("filter does not support use in client")
    }
    i, err := ib.BuildClientInterceptor(filter.Config, override)
    if err != nil {
        return nil, fmt.Errorf("error constructing filter: %v", err)
    }
    if i != nil {
        interceptors = append(interceptors, i)
    }
}
return &interceptorList{interceptors: interceptors}, nil
```

File: xds/internal/resolver/serviceconfig.go (hunk 5 above)
Comment (dfawley) [reply xds/internal/resolver/serviceconfig.go:146]: @easwars is out of the office for a couple weeks. Let's just go ahead and change it.  The only way to improve it beyond that would be to pass the actual xds enum type down to this layer, but that doesn't seem worth doing just to have a slightly better error message.

Comment (Aditya-Sood) [reply xds/internal/resolver/serviceconfig.go:146]: sure @dfawley, I'll implement it with the router check changes

@zasweq could you please give the confirmation for the newInterceptor() changes here - https://github.com/grpc/grpc-go/pull/6248#issuecomment-1582237029

**Comment by Aditya-Sood** [comment]
hi @dfawley @zasweq @easwars 
so introducing the additional `rt.actionType == xdsresource.RouteActionRoute` check is causing some existing tests to fail, because they previously didn't expect to be errored-out for failing to meet this condition

do they need to be rewritten to:
1. be of `xdsresource.RouteActionRoute` type, OR
2. expect the `"matched route does not have a supported route type"` error?

awaiting your inputs

**Comment by easwars** [comment]
I finally caught up on this set of issues/PRs. I also spoke with @zasweq. To summarize:
- The existing validation in the xdsClient to check that the last filter is a terminal filter will stay. We will not have an explicit check there to validate that the last filter *is* a router filter. Since the only supported terminal filter is the router filter, the validation in the xdsClient is good enough and does not require any changes.
- Update `newInterceptor()` method to not check for terminal filter or router filter, or ignore filters after a router filter. The change suggested in https://github.com/grpc/grpc-go/pull/6248#issuecomment-1582237029 seems about right.
- For failing tests, they should be rewritten to use `xdsresource.RouteActionRoute`
- We also need a test for the unsupported route type error

@zasweq : Please let me know if I have missed anything.

@Aditya-Sood : Please let us know if you have more questions. Thanks.

**Comment by zasweq** [comment]
That all sounds good to me :).

**Comment by easwars** [comment]
> also regarding the changes to newInterceptor(), it seems like specifically the router type filters need to be ignored and not appended to the interceptors slice

I think what we are saying is that since the xdsClient ensures that the last filter in the list is a terminal filter (and the only supported terminal filter is the `router` filter), the config selector can simply trust that the xdsClient has done its job and run through the list of filters without any validation.

**Comment by Aditya-Sood** [comment]
hi @easwars and @zasweq 

TL;DR - I need help with understanding the design/logical flow of the codebase (especially the xds resolver and router relevant parts), where can I find the low-level documentation?

I've spent more than a week trying to understand the control flow by following the error stacktraces and object definitions
With that I've been able to resolve the issue of existing test cases failing (tests have been adapted in the [above commit](https://github.com/grpc/grpc-go/pull/6248/commits/5aca643df96199349b41e5be7943e563887a45bf))

But I think I've hit a wall in terms of how much I can get done without gaining a deeper understanding of the logical flow 😬 

For instance, if I remove the following router filter check in `newInterceptor()`:
```
if router.IsRouterFilter(filter.Filter) {
    continue
}
```
then the `FaultInjection_MaxActiveFaults` test panics with a nil pointer dereference error

I spent the entire weekend trying to make sense of the error, but I'm unable to reconcile how the `fault_test.go` (source of the nil pointer dereference error) is related to the `unmarshal_lds.go` (whose "no filter implementation found for..." error is also mentioned in the stacktrace)

(PFA full error log - [TestFaultInjection_MaxActiveFaults.txt](https://github.com/grpc/grpc-go/files/11885381/TestFaultInjection_MaxActiveFaults.txt))

I have tried the [documentation files in the repo](https://github.com/grpc/grpc-go/tree/v1.56.1/Documentation) but they didn't help with understanding this 

Could you please redirect me to the relevant low-level documentation or comms channel where I can understand the logical flow? 
I'm currently out of ideas on how to resolve this test failure otherwise

Kindly let me know, thank you!

**Comment by easwars** [comment]
The following are some good docs to read to understand what we are trying to do with xDS and how we implement things in gRPC:
- Overall xDS design: https://github.com/grpc/proposal/blob/master/A27-xds-global-load-balancing.md
- Config selector design: https://github.com/grpc/proposal/blob/master/A31-xds-timeout-support-and-config-selector.md
- Fault injection design: https://github.com/grpc/proposal/blob/master/A33-Fault-Injection.md
- HTTP filters design: https://github.com/grpc/proposal/blob/master/A39-xds-http-filters.md

Some of the above docs are long reads, but they can given you a very good picture.

In the log file that you attached, I do see the following log line:
```
        fault_test.go:617: RPC error: rpc error: code = Unavailable desc = name resolver error: no filter implementation found for "type.googleapis.com/envoy.extensions.filters.http.router.v3.Router"
```
This indicates that the router package has not been imported. You need a line like this to import the `router` package which registers the `router` filter.
```golang
_ "google.golang.org/grpc/xds/internal/httpfilter/router" // Register the router filter.
```

**Comment by easwars** [comment]
Apart from what is mentioned in the docs listed above, a very high level flow would look like this:

- User application creates a `grpc.ClientConn` and uses the `xds:///` scheme in the dial target
- This results in gRPC instantiating an xDS resolver
- The xDS resolver creates an xDS client to the management server (or xDS server) to receive configuration
- The xDS resolver registers a watch for two types of resources from the xDS client, the `Listener` resource and the `RouteConfiguration` resource
  - The xDS client in turn converts these watch requests into xDS requests to the management server
  - When the management server responds with xDS resources, the xDS client takes care of unmarshaling the response, and converting it to an internal representation which is then pushed to the xDS resolver
- The xDS resolver upon receiving configuration from the xDS client creates a service config and pushes this to gRPC
  - This service config contains the LB policy to use and it also contains the config selector (which is used at RPC time to determine RPC method specific configuration)
- gRPC instantiates the LB policy specified in the service config
  - For the xDS case, the top-level LB policy will always be the cluster manager LB policy
  - There will be a tree of LB policies under this, and each of them perform a specific role
  - Put together though, they determine the configuration for the xDS cluster and the list of endpoints and also balance load across these endpoints

I know this description is very very high level, but I hope this and the above mentioned docs help you, and if you have subsequent questions, please let us know.

**Comment by Aditya-Sood** [comment]
thanks a lot @easwars! I'll try to finish these over the weekend and get back

**Comment by Aditya-Sood** [comment]
sorry for the delay in resolution, have been caught up with my day job
I've been going through the shared docs, will try to close the PR over the weekend

**Comment by Aditya-Sood** [comment]
hi @easwars and @zasweq
apologies for the delay, I had written the test case last week but wasn't fully sure that the logic of recreating it was correct, so was going through docs this week still

1) thanks to @easwars's bug fix (blank identifier import) and shared documentation, I was able to understand the high-level flow and write the new test case by changing the route config update for the new test case with a `RouteActionUnsupported` ActionType - [link to commit](https://github.com/Aditya-Sood/grpc-go/commit/cd24581cdf52d773b8d07950ff79f5b3391a0822)

2) can this PR be re-opened? if not I will create a new one

3) one query - in the `resolver.SelectConfig()` code, would it be semantically better to replace the `rt.actionType != xdsresource.RouteActionRoute` check with `rt.actionType == xdsresource.RouteActionUnsupported`, since the latter is already used to represent routing types not currently supported by gRPC?

**Comment by easwars** [comment]
> 3. one query - in the `resolver.SelectConfig()` code, would it be semantically better to replace the `rt.actionType != xdsresource.RouteActionRoute` check with `rt.actionType == xdsresource.RouteActionUnsupported`, since the latter is already used to represent routing types not currently supported by gRPC?

`resolver.SelectConfig()` happens on the client side and on that that the only supporting action type is `xdsresource.RouteActionRoute`.

File: xds/internal/resolver/serviceconfig.go [hunk 6]
```diff
@@ -142,6 +142,7 @@ type configSelector struct {
 }
 
 var errNoMatchedRouteFound = status.Errorf(codes.Unavailable, "no matched route was found")
+var errMatchedRouteTypeNotRouteActionRoute = status.Errorf(codes.Unavailable, "matched route does not have a supported route type")
```
[the remark this answers is not in the document]
Comment (easwars) [reply xds/internal/resolver/serviceconfig.go:145]: Also, the error string should be `matched route does not have a supported route action type`

File: xds/internal/resolver/xds_resolver_test.go [hunk 7]
```diff
@@ -1802,42 +1806,38 @@ func (s) TestXDSResolverHTTPFilters(t *testing.T) {
...
+				VirtualHosts: []*xdsresource.VirtualHost{
+					{
+						Domains: []string{targetStr},
+						Routes: []*xdsresource.Route{{
+							Prefix: newStringP("1"),
+							WeightedClusters: map[string]xdsresource.WeightedCluster{
+								"A": {Weight: 1},
+								"B": {Weight: 1},
+							},
+							ActionType: xdsresource.RouteActionUnsupported,
```
Comment (easwars) [inline xds/internal/resolver/xds_resolver_test.go:1830]: Maybe another similar test for action type `RouteActionNonForwardingAction`.

**Comment by Aditya-Sood** [comment]
hi @easwars, thanks for the review!
I have included the above changes in the latest commit

1. I've changed the variable name to `errUnsupportedClientRouteAction` instead of the suggested `errUnsupportedRouteAction` - I felt the latter could be confused as an error specifically for `RouteActionUnsupported` action type
However if you feel the clarification is unnecessary kindly let me know, I'll change it accordingly

2. For the "NewStream error" testcase I've removed the second route and RPC response - not sure why it was added in the first place, but the testcase passes after the change as well

3. One query: As clarified in the previous comment, the client requests will always be assigned the `RouteActionRoute` action type by the xds client ([source code snippet](https://github.com/grpc/grpc-go/blob/cd24581cdf52d773b8d07950ff79f5b3391a0822/xds/internal/xdsclient/xdsresource/unmarshal_rds.go#L299-L390))
Since that is the case, is the additional check introduced by this PR and the testcases for `RouteActionUnsupported`/`NonForwardingAction` primarily from a completeness perspective? Or is it possible to have client requests which are not of `RouteActionRoute` type?

**Comment by easwars** [comment]
> 3. One query: As clarified in the previous comment, the client requests will always be assigned the `RouteActionRoute` action type by the xds client ([source code snippet](https://github.com/grpc/grpc-go/blob/cd24581cdf52d773b8d07950ff79f5b3391a0822/xds/internal/xdsclient/xdsresource/unmarshal_rds.go#L299-L390))
>    Since that is the case, is the additional check introduced by this PR and the testcases for `RouteActionUnsupported`/`NonForwardingAction` primarily from a completeness perspective? Or is it possible to have client requests which are not of `RouteActionRoute` type?

Currently, the only supported route action type on the client is `RouteActionRoute` and the only supported route action type on the server is `NonForwardingAction`. The gRPC client and server share the same `xdsClient` which is the one which receives the resource from the management server and parses it. When parsing, the `xdsClient` does not have enough information to know whether the route configuration will be used by the client or the server (because the same application can have a server and a client). So, it will accept routes with either of these two values. And therefore we need extra validation at the client and server sides.

Yes, including test cases for `RouteActionUnsupported` and `NonForwardingAction` is for completeness.

File: xds/internal/resolver/serviceconfig.go [hunk 8]
```diff
@@ -279,12 +285,8 @@ func (cs *configSelector) newInterceptor(rt *route, cluster *routeCluster) (ires
 		return nil, nil
 	}
 	interceptors := make([]iresolver.ClientInterceptor, 0, len(cs.httpFilterConfig))
+
```
Comment (easwars) [inline xds/internal/resolver/serviceconfig.go:288]: Nit: please nix this newline.

**Review by easwars — APPROVED** [review]

**Comment by easwars** [comment]
@zasweq for second set of eyes

**Review by zasweq — APPROVED** [review]
LGTM. Doug mentioned that he wishes that the Route Type enum can be plumbed all the way from the client so we can log the exact action that was unsupported, rather than blanket it in a RouteActionUnsupported log (equivalent to what the blanket enum type which is emitted from the xDS Client). However, unfortunately, we don't emit the proto enum for Route Action, so just something to think about for the future if we want to change it. This change fixing the semantics of the error message is an improvement though :), thanks for contribution.

**Comment by Aditya-Sood** [comment]
Merged branch fixes #6259 as well
